File: MyCode/Ch10/Ex10_10.hs
--------------------------------------------------------------------
--
--   Ex10_10.hs
--
--   Kevin Baird
--
--   Aug 2007
--
--------------------------------------------------------------------

module Ex10_10 where

slope :: (Float -> Float) -> (Float -> Float)
slope f = (\x -> (f (x + 1) - f (x - 1)) / 2.0)

flat :: Float -> Float
flat   = (\x -> 1.0)

linear :: Float -> Float
linear = (\x -> x)

square :: Float -> Float
square = (\x -> x * x)

{-
map (slope flat)   [ 0.0 .. 9.0 ]
map (slope linear) [ 0.0 .. 9.0 ]
map (slope square) [ 0.0 .. 9.0 ]
-}

File: MyCode/Ch10/Ex10_10.rb
#!/usr/bin/env ruby
# Ex10_10.rb

def slope(f)
  return lambda do |x|
    (f[x+1].to_f - f[x-1].to_f) / 2.0
  end
end

procs = {
  :flat   => lambda { |x| 5 },
  :linear => lambda { |x| x },
  :square => lambda { |x| x * x },
}

0.upto(9) do |x|
  procs.keys.each do |proc_sym|
    current_proc = procs[proc_sym]
    puts "#{x}: #{current_proc.call(x)} | #{slope(current_proc).call(x)}"
  end
end
